### ISSUE TITLE
Cody: VS Code extension - Change cwd for git remote get-url origin to path from open file

### ISSUE THREAD

**Issue by marcleblanc2** [sourcegraph/cody#196]
#### Feature request description
- Multi-repo VS Code workspaces is a common use case for many of our users (opening a VS Code workspace at a parent directory with many git repos in subdirectories, ex. `~/`, or `~/git`), especially users who switch between a dozen different repos in their day-to-day work, especially users in roles in more read-heavy roles
- Cody breaks this use case because Cody depends too much on `workspaceRoot` when determining `cody.codebase` automatically, ex [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/commit/0c5b72fd46ca66e9aa27ddcdff655bf2280efcc2#diff-4dfc0d6a76ad2b610be46b871eac3413R261)
- Forcing users to open a new VS Code workspace for each repo they need to use Cody with creates way more day-to-day friction than users are willing to overcome to become daily active users of Cody
- Forcing users to configure `cody.codebase` manually in their workspace settings is a frequent cause of errors, including users configuring `cody.codebase` in their user settings scope, resulting in getting the same embeddings for any repo they open

Current state:

- We call `git remote get-url origin` with `cwd: workspaceRoot` in `getCodebaseContext` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e4c169a2153c9da57728a48cb1eeae268b97af11/-/blob/client/cody/src/chat/ChatViewProvider.ts?L903)
- 
- We call `getCodebaseContext` in `vscode.workspace.onDidChangeConfiguration` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e4c169a2153c9da57728a48cb1eeae268b97af11/-/blob/client/cody/src/chat/ChatViewProvider.ts?L120)
- We also call `getCodebaseContext` in `updateCodebaseContext` [here] https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e4c169a/-/blob/client/cody/src/chat/ChatViewProvider.ts?L434
- 
- In the default scenario, where the user doesn't have `cody.codebase` configured, `getCodebaseContext` runs `git remote get-url origin` to infer the codebase context code host / org / repo name, so that we can check if their configured Sourcegraph instance has the embeddings for this repo

Problem:
- We call `git remote get-url origin` from the current working directory of the VS Code workspace root, which assumes that the user opened the workspace in or below the root directory of a git repo
- So, when users open a workspace in, for example, a parent directory that has many git repos in many subdirectories, `git remote get-url origin` fails, so the codebase context fails to set, so users see an error message that this repo isn't indexed, which is frustrating and confusing when they know they've generated embeddings for the repo

Desired future state:
- Call `git remote get-url origin` from the directory where the file is stored that the user opened / changed to
- This would be a much more positive experience for users, as it will set the codebase context accurately more often
- This will look like we have, or may actually be close enough, to multi-repo workspaces, we're so close already
- This could be accurate enough that we could make this the default method of setting the codebase context, and only fall back to the user's configured `cody.codebase` if this method fails

TL;DR use the `vscode.window.activeTextEditor.document.uri.fsPath` file path instead of `workspaceRoot`,  when calling     `const gitCommand = spawnSync('git', ['remote', 'get-url', 'origin'], { cwd: workspaceRoot })` in `getCodebaseContext` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/cody/src/chat/ChatViewProvider.ts?L865).

Screenshot to help explain
<img width="2032" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/7050295/d1f2edb7-26c6-469b-b9ec-52bc478d5128">

On the left, I have a file open from the sourcegraph/handbook repo
On the right, I have a file open from the sourcegraph/sourcegraph repo

When I click in the file on the left, I want the Cody codebase to get updated to the handbook repo
When I click in the file on the right, I want the Cody codebase to get updated to the sourcegraph repo

We could implement a new `getActiveTextEditorFilePath()`, similar to `getWorkspaceRootPath()` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f797d27d8112891fc5238ef76b8ae3639682d627/-/blob/client/cody/src/editor/vscode-editor.ts?L16) for efficiency, so that we're only returning the `vscode.window.activeTextEditor.document.uri.fsPath` value, instead of calling `getActiveTextEditor()`, as that also returns the document content.

Then we could add a couple lines to `getCodebaseContext` to prefer the file path of the active editor, then fall back to the workspace root for situations like, there is no file open.

```typescript
async function getCodebaseContext(config: Config, rgPath: string, editor: Editor): Promise<CodebaseContext | null> {
    const client = new SourcegraphGraphQLAPIClient(config)

    // Try to get the active text editor's file path
    gitWorkingDirectory = editor.getActiveTextEditorFilePath()

    // If that fails, then the fall back is the workspace root
    if (!gitWorkingDirectory) {
 
           gitWorkingDirectory = editor.getWorkspaceRootPath()
 
           if (!gitWorkingDirectory) {
                return null
            }
    }
    const gitCommand = spawnSync('git', ['remote', 'get-url', 'origin'], { cwd: gitWorkingDirectory })
```

- We already listen for the active editor change event `vscode.window.onDidChangeActiveTextEditor` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f797d27/-/blob/client/cody/src/chat/ChatViewProvider.ts?L138:13)
- which calls `updateCodebaseContext` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f797d27d8112891fc5238ef76b8ae3639682d627/-/blob/client/cody/src/chat/ChatViewProvider.ts?L392)
- which calls `getCodebaseContext` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f797d27d8112891fc5238ef76b8ae3639682d627/-/blob/client/cody/src/chat/ChatViewProvider.ts?L865)
- which calls the `git remote get-url origin` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f797d27d8112891fc5238ef76b8ae3639682d627/-/blob/client/cody/src/chat/ChatViewProvider.ts?L865), with the `cwd: workspaceRoot`

#### Is your feature request related to a problem? If so, please describe.

Support tickets this issue would have entirely prevented:
- [13922](https://sourcegraph.zendesk.com/agent/tickets/13922)
- [13914](https://sourcegraph.zendesk.com/agent/tickets/13914)
- [13868](https://sourcegraph.zendesk.com/agent/tickets/13868)
- [13639](https://sourcegraph.zendesk.com/agent/tickets/13639)
- [13618](https://sourcegraph.zendesk.com/agent/tickets/13618)

Even with the many improvements we've made, initial setup and ongoing maintenance of the `cody.codebase` setting is still a barrier to getting started for many Cody users, so this suggestion would be a huge help in the default case for new users where they haven't configured `cody.codebase` yet, as it would automatically set the context for the user each time they open or switch to a file.

#### Additional context

Slack thread in ask-cody [here](https://sourcegraph.slack.com/archives/C04MSD3DP5L/p1684534383574339)

**Comment by sanderginn** [comment]
@marcleblanc2 I'm removing this from the EE project board because it doesn't seem to be an EE issue. Feel free to re-add if you think it should belong there (with some motivation why 😄 )

**Comment by marcleblanc2** [comment]
Was added to the product.
